### PR TITLE
[coop-gc] Make Thread.Sleep interruptable signalless

### PR DESCRIPTION
**Pull request by ludovic-henry**

### CODE REVIEW

**Comment by kumpera** [comment]
I don't like this approach, we should not be allocating memory + 2 semaphores per sleep operation.

What about having a single global cond var, do a timedwait on it.
On cancelation you'd set a flag and do broadcast.

Sleep would use a side clock and retrying to ensure it can sleep the requested time.

**Comment by ludovic-henry** [comment]
The approach has been fixed: it now uses a global cond + mutex, and will go back to waiting if it's not the current thread that has been interrupted and the time is not out.

**Comment by ludovic-henry** [comment]
build

File: mono/io-layer/wthreads.c [hunk 1]
```diff
@@ -217,6 +219,58 @@ GetCurrentThreadId (void)
...
+	mono_mutex_unlock (&sleepex_mutex);
+}
+
+static guint32
+sleepex_interruptable (guint32 ms)
+{
+	gboolean interrupted;
+	guint32 now, end;
+
+	end = mono_msec_ticks () + ms;
```
Comment (ludovic-henry) [inline mono/io-layer/wthreads.c]: check for overflow

File: mono/io-layer/wthreads.c [hunk 2]
```diff
@@ -217,6 +219,68 @@ GetCurrentThreadId (void)
...
+static void
+sleepex_interrupt (gpointer data)
+{
+	mono_mutex_lock (&sleepex_mutex);
+	mono_cond_broadcast (&sleepex_cond);
+	mono_mutex_unlock (&sleepex_mutex);
+}
+
+static guint32
+sleepex_interruptable (guint32 ms)
```
Comment (kumpera) [inline mono/io-layer/wthreads.c]: This is the wrong way to handle time wrap around. Ticks wrap around is a lot more common than you'd think.
It happens every 1193 hours, or every 49 days.

The condition to check is not 'now < end' but 'now - start >= ms'.

now - start works fine even in the case of overflow due to how unsigned 2-complement math works.

Let's see how it works with a 1 byte unsigned value. start is 0xFE and we have advanced 1, 2 and 5 (values of now).

0xFE + 1 => 0xFF
0xFE + 2 => 0x00
0xFE + 5 => 0x03

Now let's do the reverse operation (values of now - start):

0xFF - 0xFE => 1
0x00 - 0xFE => 2
0x03 - 0xFE => 5

This ensures that as long as ms is smaller that MAXUINT, we can handle overflow just fine.

**Comment by kumpera** [comment]
Besides the wrong handling of overflow, maybe we could special case small values of sleep and not do the heavy lifting?

This is particularly important for the case of Sleep(0) which is used as a replacement for sched_yield ().

Maybe up to 10ms?

**Comment by ludovic-henry** [comment]
Ok for the overflow.

For the special case for 0, it's already done in SleepEx. As for special case for values of up to 10ms (for example), that would add thread interrupt and abort up to 10ms of latency, which is not necessarily a good behaviour IMO.

**Comment by ludovic-henry** [comment]
build

**Pull request closed without merging**